### ISSUE TITLE
Auto-Resistance not working with Nordictrack GX4.5 bike

### ISSUE THREAD

**Issue by AndyMac1972** [cagnulein/qdomyos-zwift#2233]
**Describe the bug**
Firstly...awesome app thanks! All works perfectly except the auto-resistance is not working at all with my Nordictrack GX4.5 bike.

I can manually adjust the resistance on the bike and this adjusts the watts calculation in Zwift...so the info seems to be getting from the bike to the app, but not the other way around. In the QZ app window the "-" and "+" do not alter the resistance, but also does not show the current resistance setting as shown on the bike.

**To Reproduce**
Not applicable

**Expected behavior**
Expect that resistance will alter based on incline in free ride and that in ERG mode the resistance will alter to hit target watts based on cadence

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**
 - Device: Samsung S21
 - OS: Android
 - Version 13

**Append a debug log**
Please let me know if you need a debug log. 

 Follow this guide https://github.com/cagnulein/qdomyos-zwift/wiki/How-do-i-get-the-debug-log-in-case-something-doesn't-work%3F
  
  
**Additional context**

**Comment by cagnulein** [comment]
Hi! Thank you! In order to add auto resistance i need a log from the ifit
app following this
https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/20_supported_devices_and_applications.md#android
device

You have to change the resistance from the app directly


Let me know if you have any questions

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 15 mar 2024 alle 21:21 AndyMac1972 ***@***.***>
ha scritto:

**Comment by AndyMac1972** [comment]
Thanks for the super-fast response. I will get this back to you later today.

**Comment by AndyMac1972** [comment]
Hi Roberto

Here is the bug file...hope I've done it correctly. Let me know if you need
anything else.

Thanks again!
Andy

**Comment by cagnulein** [comment]
Hi, I guess you forgot the attachment. You have to upload it to github directly otherwise I can't see it

**Comment by AndyMac1972** [comment]
[bcm_bt_ewp.log](https://github.com/cagnulein/qdomyos-zwift/files/14622728/bcm_bt_ewp.log)
[btsnoop_hci.log](https://github.com/cagnulein/qdomyos-zwift/files/14622729/btsnoop_hci.log)

Yep...novice in Github sorry. I replied from my email inbox. Let me try again!

Thanks
Andy

**Comment by cagnulein** [comment]
Ok I will check this next week, thanks!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 16 mar 2024 alle 08:04 AndyMac1972 ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
i was briefly checking your log and i saw that the "Cycle Trainer 300 CI" profile is very similar to your bike. Please, try to enable this setting in the proform setting and let me know. leave this only one enable there and restart qz
Let me know!

**Comment by AndyMac1972** [comment]
I only have "Cycle Trainer 400" in the list? Should I try this?

**Comment by cagnulein** [comment]
nope, join the beta from here https://play.google.com/apps/testing/org.cagnulen.qdomyoszwift
and update the app if you didn't already yet
you should see version 2.16.47 
then you will see also the "Cycle Trainer 300 CI"
Let me know

**Comment by AndyMac1972** [comment]
Thanks Roberto

I tried this but auto resistance still not working. Do you want another
debug file created?

Cheers
Andy

**Comment by cagnulein** [comment]
nope, i will check first the hci snoof deeply. it was just a quick attempt.

i will do in the next days

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 17 mar 2024 alle 02:04 AndyMac1972 ***@***.***>
ha scritto:

**Comment by AndyMac1972** [comment]
Thanks Roberto.

No rush!

**Comment by cagnulein** [comment]
> [bcm_bt_ewp.log](https://github.com/cagnulein/qdomyos-zwift/files/14622728/bcm_bt_ewp.log) [btsnoop_hci.log](https://github.com/cagnulein/qdomyos-zwift/files/14622729/btsnoop_hci.log)
> 
> Yep...novice in Github sorry. I replied from my email inbox. Let me try again!
> 
> Thanks Andy

@AndyMac1972 did you change the resistance from the ifit app directly? I don't see them in the hci snoof.

**Comment by AndyMac1972** [comment]
I'm not using the iFit app...I'm using Zwift. The reason I got your app was so i didn't have to use iFit (as I think they are crooks!). I used to subscribe, but got rid of it a long time ago, but it worked no problem.
Sorry i didn't make this clear earlier.

**Comment by cagnulein** [comment]
no i mean, I need a hci snoof where you change the resistance directly from ifit so i can sniff the protocol and add auto resistance to qz so you can use it with zwift.

**Comment by AndyMac1972** [comment]
gotcha. so i can just do this from manual mode...don't need to subscribe? I will try again!!! Thanks for your patience

**Comment by cagnulein** [comment]
exaclty, manual mode is enough!

**Comment by AndyMac1972** [comment]
Hi Roberto
Here goes v2...hopefully got it right this time.
[bcm_bt_ewp.log](https://github.com/cagnulein/qdomyos-zwift/files/14705260/bcm_bt_ewp.log)
[btsnoop_hci.log](https://github.com/cagnulein/qdomyos-zwift/files/14705261/btsnoop_hci.log)

**Comment by cagnulein** [comment]
Thanks I will check it tomorrow

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 21 mar 2024 alle 19:04 AndyMac1972 ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
@AndyMac1972 just to be sure, did you use the ifit app from the same phone where are you collecting the log? because i'm seeing the same data that I send for the proform_cycle_trainer_300_ci so i would like to be sure about this point. Maybe there is a small difference but before checking deeply I would like to be sure about this point :)
Thanks

**Comment by AndyMac1972** [comment]
Yep. iFit was running on the same phone.

**Comment by cagnulein** [comment]
ok so i will put my inspector gadget's pants now :D

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 22 mar 2024 alle ore 09:32 AndyMac1972 <
***@***.***> ha scritto:

**Comment by AndyMac1972** [comment]
nice one! good luck! trying not to picture that in my mind!

**Comment by cagnulein** [comment]
ok i guess i got it. the bike has only 10 resistance levels? or you just
changed to 10 level in the log?


Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 22 mar 2024 alle ore 09:59 AndyMac1972 <
***@***.***> ha scritto:

**Comment by AndyMac1972** [comment]
Sorry no. It has 25 levels...I only took it up to 10 in the bug file. Do you need me to re-run it through the entire range?

**Comment by cagnulein** [comment]
yes please because each resistance level has a different code that's why
the current implementation doesn't work

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 22 mar 2024 alle 20:17 AndyMac1972 ***@***.***>
ha scritto:

**Comment by AndyMac1972** [comment]
Okay. Will get that done today. Thanks

**Comment by cagnulein** [comment]
Great.
i will be back on monday :)

Il giorno ven 22 mar 2024 alle 20:23 AndyMac1972 ***@***.***>
ha scritto:

**Comment by AndyMac1972** [comment]
[bcm_bt_ewp.log](https://github.com/cagnulein/qdomyos-zwift/files/14740351/bcm_bt_ewp.log)
[btsnoop_hci.log](https://github.com/cagnulein/qdomyos-zwift/files/14740352/btsnoop_hci.log)

**Comment by cagnulein** [comment]
hi @AndyMac1972 in this log i see the communication with the bike but i don't see any resistance changing from the app itself. Did you change all the levels from the app directly? You must in this way in order to collect all the info that I need. From lvl 1 to lvl max and lvl 1 back (keep 5 seconds between each step to be sure) directly from the app

Let me know!

**Comment by AndyMac1972** [comment]
[bcm_bt_ewp.log](https://github.com/cagnulein/qdomyos-zwift/files/14808397/bcm_bt_ewp.log)
[btsnoop_hci.log](https://github.com/cagnulein/qdomyos-zwift/files/14808398/btsnoop_hci.log)

Hopefully this time! Sorry it's taken a while to reply...busy week!

**Comment by cagnulein** [comment]
thanks, i will check it after easter days! happy easter!

**Comment by AndyMac1972** [comment]
Thanks...you too!

**Comment by cagnulein** [comment]
@AndyMac1972 done! i did a lot of code for this so maybe we need some versions to stabilize it.
So download the new version from here https://github.com/cagnulein/qdomyos-zwift/actions/runs/8519171686 and enable the new gx4.5 setting under the proform settings

Google is very slow in these days, reviewing the app on the store, even the beta ones. For this reason, in order to check right now if the patch that I did is right, you can use the android build in the link (check at the bottom of the page in about 1 hour). In order to test it you need first to remove the current QZ version on your phone, install the one that you downloaded.
This version is a 15 minutes trial (you can restart it as many times as you want) and I can unlock it if you want, send me an email to roberto.viola83@gmail.com mentioning this ticket and that you are on android.
Let me know if you have any questions

**Comment by AndyMac1972** [comment]
Hi Roberto. I have updated the app and trialed it. The resistance reading in the app window is now matching the resistance setting on the bike, however the + and - in the app do not adjust it. Also, when in a free ride, the resistance does not change depending on the incline. I have only had a chance to run 1 15minute cycle, but will have time over the weekend if you can suggest what i should check next. 

Thanks again for all your help!

**Comment by AndyMac1972** [comment]
Hi Roberto...this file is very small, but was the only one in the BT directory? Let me know if I need to run it again. I ran the trial for 15 minutes across a couple of different rides. Let me know if you want me to have another try.
[bcm_bt_ewp.log](https://github.com/cagnulein/qdomyos-zwift/files/14895012/bcm_bt_ewp.log)

**Comment by cagnulein** [comment]
hi, no i mean i need a debug log from qz not a hci a snoof file. the one
generated following this
https://github.com/cagnulein/qdomyos-zwift/wiki/How-do-i-get-the-debug-log-in-case-something-doesn't-work%3F

Let me know if something is unclear

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 6 apr 2024 alle 23:21 AndyMac1972 ***@***.***>
ha scritto:
